apps/frontend/src/components/snapshots/snapshot-form-dialog.tsx

```diff
@@ -0,0 +1,278 @@
+import { Button } from '@/components/ui/button';
+import DatePicker from '@/components/ui/datepicker';
+import {
+  Dialog,
+  DialogContent,
+  DialogHeader,
+  DialogTitle,
+  DialogTrigger,
+} from '@/components/ui/dialog';
+import { useContext, useEffect, useState } from 'react';
+import { Observation, Snapshot } from 'types/database_types';
+import { z } from 'zod';
+import { useUser } from '../../hooks/useUser';
+import ObservationsSection from '../observations/observations-section';
+import HistoricalSnapshotContext from './historical-snapshot-context';
+import LatestSnapshotContext from './latest-snapshot-context';
+
+const snapshotSchema = z.object({
+  snapshotID: z.number().optional(),
+  dateCreated: z.date(),
+  patchID: z.string(),
+  notes: z.string(),
+  userID: z.string().optional(),
+});
+
+export default function SnapshotForm({
+  newSnapshot,
+  patchID,
+  snapshotTemplate,
+  observationsTemplate,
+}: {
+  newSnapshot: boolean;
+  patchID: string;
+  snapshotTemplate?: Snapshot | null;
+  observationsTemplate?: Observation[] | null;
+}) {
+  const [open, setOpen] = useState(false);
+  const [notes, setNotes] = useState<string>('');
+  const [date, setDate] = useState<Date | null>(null);
+  const [observations, setObservations] = useState<Observation[]>([]);
+  const { user } = useUser();
+  const { fetchLatestSnapshot } = useContext(LatestSnapshotContext);
+  const { fetchHistoricalSnapshotMetadata } = useContext(HistoricalSnapshotContext);
+
+  useEffect(() => {
+    if (snapshotTemplate && newSnapshot === false) {
+      setNotes(snapshotTemplate.notes || '');
+      setDate(new Date(snapshotTemplate.dateCreated)); // Ensure the date is in the correct format, fallback to empty string if null
+    } else {
+      setNotes('');
+      setDate(null);
+    }
+  }, [snapshotTemplate, newSnapshot]);
+
+  useEffect(() => {
+    if (observationsTemplate && newSnapshot === false) {
+      setObservations(observationsTemplate);
+    } else {
+      setObservations([]);
+    }
+  }, [observationsTemplate, newSnapshot]);
+
+  const duplicateLatestData = () => {
+    if (!snapshotTemplate) {
+      console.error('No snapshot template to duplicate from');
+      return;
+    }
+    if (!observationsTemplate) {
+      console.error('No observations template to duplicate from');
+      return;
+    }
+    setNotes(snapshotTemplate.notes || '');
+    setDate(new Date(snapshotTemplate.dateCreated));
+    setObservations(observationsTemplate);
+  };
+
+  async function onSubmit() {
+    if (!date) {
+      alert('Please select a date for the snapshot.');
+      return;
+    }
+    if (notes.trim() === '') {
+      alert('Please enter notes for the snapshot.');
+      return;
+    }
+    if (!user || !user.id) {
+      console.error('User is not authenticated or missing user ID.');
+      alert('You must be logged in to submit a snapshot. Please log in and try again.');
```

Comment:
Use toasts rather than alerts. Generally, we should never need to use alerts at all

Reply:
In any case, a person that isn't logged in or is just a regular user should not be able to enter this flow to begin with